torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -386,6 +406,96 @@ auto handle_torch_function_no_python_arg_parser(
+    auto found_proxy_or_fake = false;
+    auto is_proxy_mode = false;
+    auto is_fake_mode = false;
+    c10::optional<py::object> torch_dispatch;
+
+    // Try proxy mode
+    auto maybe_proxy_mode = c10::impl::TorchDispatchModeTLS::get_proxy_mode();
+    if (maybe_proxy_mode != c10::nullopt) {
+      mode_obj = py::reinterpret_borrow<py::object>(
+          (*maybe_proxy_mode)->ptr(getPyInterpreter()));
+      is_proxy_mode = true;
+      found_proxy_or_fake = true;
+    }
+
+    // Try fake mode
+    if (!found_proxy_or_fake) {
+      auto maybe_fake_mode = c10::impl::TorchDispatchModeTLS::get_fake_mode();
+      if (maybe_fake_mode != c10::nullopt) {
+        mode_obj = py::reinterpret_borrow<py::object>(
+            (*maybe_fake_mode)->ptr(getPyInterpreter()));
+        is_fake_mode = true;
+        found_proxy_or_fake = true;
+      }
+    }
+
+    // Looks for fake tensor args
+    if (!found_proxy_or_fake) {
+      for (auto& arg : overloaded_fake_args) {
+        torch_dispatch =
+            PyObject_FastGetAttrString(arg.ptr(), torch_function_name_str);
+        found_proxy_or_fake = true;
+      }
+    }
+    if (found_proxy_or_fake) {
+      // We successfully found a proxy / fake mode or argument to dispatch on
+      // Sigh, there are a few differences on how we need do call back into
+      // python depending on whether our object is a mode or a subclass.
+      if (is_proxy_mode || is_fake_mode) {
+        torch_dispatch_mode::ProxyOrFakeModeGuard guard(
+            /*is_proxy=*/is_proxy_mode);
+        auto mode_obj = py::reinterpret_borrow<py::object>(
+            guard.get_cur_mode()->ptr(getPyInterpreter()));
+        if (kwargs == nullptr) {
+          ret = py::reinterpret_steal<py::object>(PyObject_CallMethod(
+              mode_obj.ptr(),
+              torch_function_name_str,
+              "OOO",
+              torch_api_function,
+              py_types.ptr(),
+              args));
+        } else {
+          ret = py::reinterpret_steal<py::object>(PyObject_CallMethod(
+              mode_obj.ptr(),
+              torch_function_name_str,
+              "OOOO",
+              torch_api_function,
+              py_types.ptr(),
+              args,
+              kwargs));
+        }
+      } else {
+        // We must have found a FakeTensor subclass (with a __torch_dispatch__
+        // field)
+        if (torch_dispatch == c10::nullopt) {
+          TORCH_INTERNAL_ASSERT(0);
+        }
+        if (!(*torch_dispatch)) {
+          TORCH_INTERNAL_ASSERT(0);
+        }
+        ret = py::reinterpret_steal<py::object>(PyObject_CallFunctionObjArgs(
+            torch_dispatch->ptr(),
+            torch_api_function,
+            py_types.ptr(),
+            args,
+            kwargs,
+            NULL));
+      }
+
+      TORCH_INTERNAL_ASSERT(py_types.ptr() != nullptr);
+      TORCH_INTERNAL_ASSERT(args != nullptr);
+      if (ret.ptr() == nullptr) {
+        throw python_error();
+      }
```

Comment:
A refactor to reduce code duplication here would be mucho appreciated

Reply:
yep. Let me think more about your `__torch_fake_dispatch__` suggestion - if that doesn't work out, at the very least I'll refactor the code here.